File: 2023/day1/main.py
from typing import Optional, Iterable, Callable

Parser = Callable[[str], int]


def solution(input: Iterable[str], parser: Parser) -> int:
    numbers = (parser(line) for line in input)
    return sum(numbers)


# Part 1
def parse_number(x: str) -> int:
    numbers = [int(c) for c in x if c.isnumeric()]
    return 10 * numbers[0] + numbers[-1]


# Part 2
WORDS = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]


def check_word(s: str, word: str, index: int) -> bool:
    end = index + len(word)
    return end <= len(s) and s[index:end] == word


def check_words(s: str, index) -> Optional[int]:
    for i, word in enumerate(WORDS):
        if check_word(s, word, index):
            return i + 1
    return None


def parse_number2(x: str) -> int:
    numbers = []
    for index, v in enumerate(x):
        if v.isnumeric():
            numbers.append(int(v))
            continue
        value = check_words(x, index)
        if value:
            numbers.append(value)
    return 10 * numbers[0] + numbers[-1]


def main() -> None:
    with open("input.txt", "r") as f:
        input = f.readlines()

    part1 = solution(input, parse_number)
    print(f"Part 1: {part1}")

    part2 = solution(input, parse_number2)
    print(f"Part 2: {part2}")


if __name__ == "__main__":
    main()

File: 2023/day1/src/main.rs
use std::fs::read_to_string;

fn read_input(filename: &str) -> Vec<String> {
    read_to_string(filename)
        .unwrap()
        .lines()
        .map(String::from)
        .collect()
}

fn solution<F>(input: &[String], parser: F) -> u32
where
    F: Fn(&str) -> u32,
{
    input.iter().map(|x| parser(x)).sum()
}

fn main() {
    let input = read_input("input.txt");

    let part1: u32 = solution(&input, parser_1);
    println!("{}", part1);

    let part2: u32 = solution(&input, parser_2);
    println!("{}", part2);
}

// Part 1
fn parser_1(line: &str) -> u32 {
    let mut digits = line.chars().flat_map(|x| x.to_digit(10));
    let first_digit = digits.next();
    let last_digit = digits.last();
    match (first_digit, last_digit) {
        (Some(f), Some(l)) => 10 * f + l,
        (Some(f), None) => 10 * f + f,
        _ => panic!("Unable to parse line: {}", line),
    }
}

// Part 2
const DIGITS: [&str; 9] = [
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
];

fn check_word(line: &str, index: usize) -> Option<usize> {
    for (i, v) in DIGITS.into_iter().enumerate() {
        let end = index + v.len();
        if end <= line.len() && &line[index..end] == v {
            return Some(i + 1);
        }
    }
    None
}

fn check_digit(line: &str, index: usize) -> Option<usize> {
    (line[index..(index + 1)]).parse::<usize>().ok()
}

fn parse_at_index(line: &str, index: usize) -> Option<usize> {
    if let Some(x) = check_digit(line, index) {
        return Some(x);
    }
    check_word(line, index)
}

fn parser_2(line: &str) -> u32 {
    let n = line.len();
    let mut digits = (0..n).flat_map(|i| parse_at_index(line, i));
    let first_digit = digits.next();
    let last_digit = digits.last();
    let out = match (first_digit, last_digit) {
        (Some(f), Some(l)) => 10 * f + l,
        (Some(f), None) => 10 * f + f,
        _ => panic!("Unable to parse line: {}", line),
    };
    out.try_into().unwrap()
}
